src/conversions/std/num.rs

```diff
@@ -106,15 +107,19 @@ macro_rules! int_fits_c_long {
 
         impl<'source> FromPyObject<'source> for $rust_type {
             fn extract(obj: &'source PyAny) -> PyResult<Self> {
-                let ptr = obj.as_ptr();
-                let val = unsafe {
-                    let num = ffi::PyNumber_Index(ptr);
-                    if num.is_null() {
-                        Err(PyErr::fetch(obj.py()))
-                    } else {
-                        let val = err_if_invalid_value(obj.py(), -1, ffi::PyLong_AsLong(num));
-                        ffi::Py_DECREF(num);
-                        val
+                // fast path - checking for subclass of `int` just checks a bit in the type object
+                let val = if let Ok(long) = obj.downcast::<PyLong>() {
```

Comment:
Should this be `downcast_exact`? Can an `int` subclass override the behavior?

Reply:
From [the docs](https://docs.python.org/3/c-api/long.html#c.PyLong_AsLong)

> Return a `C long` representation of `obj`. If `obj` is not an instance of [`PyLongObject`](https://docs.python.org/3/c-api/long.html#c.PyLongObject), first call its [`__index__()`](https://docs.python.org/3/reference/datamodel.html#object.__index__) method (if present) to convert it to a [`PyLongObject`](https://docs.python.org/3/c-api/long.html#c.PyLongObject).

So I would think any subclass of `int` is okay.

I could also use `is_exact_instance_of` which I tried and has the exact same performance as this.

Reply:
To me this reads like we could always avoid calling `PyNumber_Index` because `PyLong_AsLong` will do that anyway if necessary?

Reply:
Which does indeed seem to be the case, i.e.

https://github.com/python/cpython/blob/77b45fa6d0b8c0c14657b5117b21a3f3f2ce97d8/Objects/longobject.c#L480

So we really only need the fast path?

Reply:
That matches my tests - e.g. nothing in pydantic-core's tests fails if you remove the slow path.

I assume the reason the code currently has the`PyNumber_Index` path was that the docs sentence is ambiguous - either meaning:
* "If obj is not an instance of `PyLongObject`, **you should** first call its `__index__()`..."
* or, "If obj is not an instance of `PyLongObject`, **this method will** first call its `__index__()`..." - which is I think what they meant

Should I remove the slow path from here, `BigInt` and anywhere else I find it?

Reply:
Personally, I'd say yes. We probably want to do a full CI build though to see if the `PyLong_AsLong` was implemented differently in older versions of CPython, i.e. we should as the `CI-build-full` label after the reduce PR is green.

Reply:
makes sense, it looks like even [in 3.8](https://github.com/python/cpython/blob/v3.8.18/Objects/longobject.c#L496) there was similar logic albeit with a different name.